velox/exec/Spill.h

```diff
@@ -0,0 +1,340 @@
+  // size of a single file in the file set. 'pool' and 'mappedMemory' are used
+  // for buffering and constructing the result data read from 'this'.
+  FileList(
+      TypePtr type,
+      const std::string path,
+      uint64_t targetBatchSize,
+      uint64_t targetFileSize,
+      memory::MemoryPool& pool,
+      memory::MappedMemory* mappedMemory)
+      : type_(type),
+        path_(path),
+        targetBatchSize_(targetBatchSize),
+        targetFileSize_(targetFileSize),
+        pool_(pool),
+        mappedMemory_(mappedMemory) {}
+
+  // Adds 'rows' for the positions in 'indices' into 'this'. The indices
+  // must produce a view where the rows are sorted if sorting is desired.
+  // Consecutive calls must have sorted data so that the first row of the
+  // next call is not less than the last row of the previous call.
+  void write(RowVectorPtr rows, const folly::Range<IndexRange*> indices);
+  std::vector<std::unique_ptr<SpillFile>> files() {
+    return std::move(files_);
+  }
+
+  // Closes one file of output. Subsequent calls to 'write' start a
+  // different sorted run and must be ordered between themselves but not
+  // with respect to calls before flush(). If 'close' is true, write must
+  // not be called and the file set can only be read.
+  void flush(bool close);
+
+ private:
+  const TypePtr type_;
+  const std::string path_;
+  const uint64_t targetBatchSize_;
+  const uint64_t targetFileSize_;
+  memory::MemoryPool& pool_;
+  memory::MappedMemory* const mappedMemory_;
+  std::unique_ptr<VectorStreamGroup> batch_;
+  bool isOpen_ = false;
+  std::vector<std::unique_ptr<SpillFile>> files_;
+};
+
+// Represents all spilled data of an operator, e.g. order by or group
+// by. This has one FileList per partition of spill data.
+class SpillState {
+ public:
+  // Constructs a SpillState. 'type' is the content RowType. 'path' is the file
+  // system path prefix. 'bits' is the hash bit field for partitioning data
+  // between files. This also gives the maximum number of partitions.
+  // 'targetFileSize' is the target size of a single file. 'targetBatchSize is
+  // the target number of rows in a single RowVector written to a spill file.
+  // 'pool' and 'mappedMemory' own the memory for state and results.
+  SpillState(
+      TypePtr type,
+      const std::string& path,
+      HashBitField bits,
+      uint64_t targetFileSize,
+      uint64_t targetBatchSize,
+      memory::MemoryPool& pool,
+      memory::MappedMemory* mappedMemory)
+      : type_(type),
+        path_(path),
+        bits_(bits),
+        fieldMask_(((1UL << (bits_.end - bits_.begin))) - 1),
+        targetFileSize_(targetFileSize),
+        targetBatchSize_(targetBatchSize),
+        pool_(pool),
+        mappedMemory_(mappedMemory) {}
+
+  // Returns which spill partition 'hash' falls. Returns -1 if the
+  // partition of 'hash' is has not been started.
+  int32_t way(uint64_t hash) {
```

Comment:
- "way" is a new term that's not used anywhere else and it non-intuitive. Any particular reason not to use "partition"?

- This method can be const, right?